includes/functions.php

```diff
@@ -391,6 +395,247 @@ function pmpro_loadTemplate( $page_name = null, $where = 'local', $type = 'pages
+		array( PMPRO_DIR => array( 'type' => 'pages', 'page' => 'account' ) ),
+		array( PMPRO_DIR => array( 'type' => 'pages', 'page' => 'billing' ) ),
+		array( PMPRO_DIR => array( 'type' => 'pages', 'page' => 'cancel' ) ),
+		array( PMPRO_DIR => array( 'type' => 'pages', 'page' => 'checkout' ) ),
+		array( PMPRO_DIR => array( 'type' => 'pages', 'page' => 'confirmation' ) ),
+		array( PMPRO_DIR => array( 'type' => 'pages', 'page' => 'invoice' ) ),
+		array( PMPRO_DIR => array( 'type' => 'pages', 'page' => 'levels' ) ),
+		array( PMPRO_DIR => array( 'type' => 'pages', 'page' => 'login' ) ),
+		array( PMPRO_DIR => array( 'type' => 'pages', 'page' => 'member_profile_edit' ) )
+	);
+
+	/**
+	 * Filter the current array of base templates used to compare template versions
+	 *
+	 * @since TBD
+	 *
+	 * @param array $core_page_templates The array of core templates we compare with overridden template versions
+	 */
+	$pages = apply_filters( 'pmpro_base_template_versions', $core_page_templates );
+
+	$template_versions = array();
+
+	foreach( $pages as $page ) {
+
+		foreach( $page as $location => $template_data ) {
+			
+			if( !empty( $template_data['ext'] ) ) {
+				//This will add support for things like email templates
+				$ext = $template_data['ext']; 
+			} else {
+				$ext = 'php';
+			}
+
+			$template = $location . "/".$template_data['type']."/".$template_data['page'].".{$ext}";
+
+			$template_header = pmpro_get_template_version( $template );
+
+			$template_versions[$template_data['page']] = $template_header['version'];
+		}
+
+	}
+
+	pmpro_setOption( 'template_versions', maybe_serialize( $template_versions ) );
+
+	return $template_versions;
+
+}
+
+/**
+ * Get a specific template version
+ *
+ * @param null   $pagename - Name of the template 
+ *
+ * @since TBD
+ *
+ * @return string - Version number
+ */
+function pmpro_get_used_template_version( $pagename ) {
+
+	return pmpro_getOption( 'template_version_'.$pagename );
+
+}
+
+/**
+ * Compares all base template versions with any overridden templates
+ *
+ * @since TBD
+ *
+ * @return array - An array of templates that are out of date compared to the base templates
+ */
+function pmpro_compare_template_versions() {
+
+	$templates = pmpro_get_template_versions();
+
+	$affected_templates = array();
+
+	if( !empty( $templates ) ) {
+
+		foreach( $templates as $template_slug => $version ) {
+
+			$actual_version = $version;
+
+			$used_version = pmpro_get_used_template_version( $template_slug );
+
+			if( !empty( $used_version ) ){
+				//Only compare if we have a used version number
+				$compared = version_compare( floatval( $used_version ), floatval( $actual_version ), "<");
+
+				if( $compared ) {
+					$affected_templates[$template_slug] = array( 'our_version' => $actual_version, 'your_version' => $used_version );
+				} else {
+					$affected_templates[$template_slug] = array( 'our_version' => $actual_version, 'your_version' => __( 'Unknown', 'paid-memberships-pro' ) );
+				}
+			}
```

Comment:
I was running into some issues testing around here. Even without using any custom page templates, I was still getting errors shown in the dashboard that I need to update. This is because we never check if `the actual version === the used version`.

I think the following logic would make sense here:
1. If there is not an $actual_version, show a warning that the custom template may be outdated and show "unknown" as the used version number
2. If $actual_version > $used_version, show a warning that they are using an outdated template and show version numbers
3. Otherwise,  template is ok

Reply:
Could we update the options in the upgradecheck.php file? With this release, and then any time we increment a template version, we'd have to do that again.

Reply:
I'm worried about showing a warning that could confuse the site owner.

I think I would rather detect the case where there is no actual_version and just wait until the template gets loaded. I think sometimes actual_version would be the old version until the template is loaded, and in that case we wouldn't show the warning either until the template got loaded.

I would prefer to just wait until the template gets loaded to detect the issue and "queue up" the notice than show a warning.